### PR TITLE
feat: Introduce the asm feature library that supports lazy loading. (GH-10047)

### LINKED ISSUE #10047: Introduce ASM feature library
As part of the initiative to move features in separate and optional libraries, we like to move ASM out of the core and component library.  As a result, the ASM feature should be optional, lazily loaded (code split). 

Ideally we introduce this change as an optional change during a minor release, but there's a change that we can only do this with breaking changes.

Requires these to be merged in this PR:
- [x] Address pr review comments
- [x] Enable strict mode #10697
- [x] Solve agent session issue #10854 
- [x] Fix ASM e2e #11118
- [x] Enable Schematics? No, this will be done in a generic way for all libs.

### PR DESCRIPTION
**Pull request by plabadie**
closes #10047

### CODE REVIEW

File: feature-libs/asm/root/services/asm-enabler.service.ts [hunk 1]
```diff
@@ -65,6 +81,9 @@ export class AsmEnablerService {
    * Adds the ASM UI by using the `cx-storefront` outlet.
    */
   protected addUi(): void {
-    this.launchDialogService.launch(LAUNCH_CALLER.ASM);
+    this.lazyModule
```
Comment (dunqan) [inline feature-libs/asm/root/services/asm-enabler.service.ts]: We should make sure that the module instance is resolved only once, I'm thinking about wrapping it into observable, for example `asmModuleInstance$` with `shareReplay`

File: feature-libs/asm/tsconfig.lib.json [hunk 2]
```diff
@@ -0,0 +1,29 @@
+/* To learn more about this file see: https://angular.io/config/tsconfig. */
+{
+  "extends": "../../tsconfig.json",
+  "compilerOptions": {
```
Comment (dunqan) [inline feature-libs/asm/tsconfig.lib.json:3]: Can we also enable strict mode for this library?

File: feature-libs/asm/root/services/asm-enabler.service.ts [hunk 3]
```diff
@@ -14,10 +20,25 @@ import { ASM_ENABLED_LOCAL_STORAGE_KEY } from '../asm-constants';
   providedIn: 'root',
 })
 export class AsmEnablerService {
+  private asmModuleInstance$: Observable<
+    NgModuleRef<any>
+  > = this.configInitializer.getStable('featureModules').pipe(
+    map((config: CmsConfig) => config.featureModules ?? {}),
+    switchMap((featureModulesConfig) =>
+      this.lazyModule.resolveModuleInstance(
```
Comment (dunqan) [inline feature-libs/asm/root/services/asm-enabler.service.ts]: Just realized that this solution will only work for lazy loaded asm, it should also work for eager loading ootb, the solution here could be checking, if `eatureModulesConfig['asm']?.module` is defined, and if not, just return `of(undefined)` in the `swtichMap`.

File: feature-libs/asm/tsconfig.lib.json [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    "experimentalDecorators": true,
+    "importHelpers": true,
+    "types": [],
+    "lib": ["es2020", "dom"],
+    "paths": {
+      "@spartacus/core": ["dist/core"],
+      "@spartacus/storefront": ["dist/storefrontlib"]
+    }
+  },
+  "angularCompilerOptions": {
```
Comment (dunqan) [inline feature-libs/asm/tsconfig.lib.json:21]: Can we also set
`strictTemplates` to true and check if we can keep it?

Comment (dunqan) [reply feature-libs/asm/tsconfig.lib.json:21]: Just an idea, maybe we can create a new example app with the --strict flag as described in https://angular.io/guide/strict-mode and use it as a guide to improve our tsconfig stricness?

File: projects/core/src/asm/config/asm-config.ts [hunk 5]
```diff
@@ -1,7 +1,10 @@
-import { OccConfig } from '../../occ/config/occ-config';
 import { Injectable } from '@angular/core';
 import { Config } from '../../config/config-tokens';
+import { OccConfig } from '../../occ/config/occ-config';
 
+/**
+ * @deprecated since 3.0, use asm lib instead
```
Comment (dunqan) [inline projects/core/src/asm/config/asm-config.ts]: I believe it should be 3.2 finally

File: projects/assets/src/translations/en/index.ts [hunk 6]
```diff
@@ -11,7 +10,6 @@ import { user } from './user';
 
 export const en = {
   address,
-  asm,
```
Comment (dunqan) [inline projects/assets/src/translations/en/index.ts]: We have to keep it, but mark it with some todo, so we will remove it in 4.0

File: projects/storefrontlib/src/cms-components/cms-lib.module.ts [hunk 7]
```diff
@@ -63,7 +62,7 @@ import { WishListModule } from './wish-list/wish-list.module';
 @NgModule({
   imports: [
     AnonymousConsentManagementBannerModule,
-    AsmModule,
+    //AsmModule,
```
Comment (dunqan) [inline projects/storefrontlib/src/cms-components/cms-lib.module.ts]: We have to bring it back (just to keep track of it)

File: projects/storefrontlib/src/recipes/storefront.module.ts [hunk 8]
```diff
@@ -28,9 +27,7 @@ import { StorefrontFoundationModule } from './storefront-foundation.module';
     StoreModule.forRoot({}),
     EffectsModule.forRoot([]),
 
-    // ASM module must be imported before the `AuthModule (which is imported in `StorefrontFoundationModule`)
-    // since we might have conflicting interceptor logic. See #5461.
-    AsmModule,
+    //AsmModule,
```
Comment (dunqan) [inline projects/storefrontlib/src/recipes/storefront.module.ts]: We have to bring it back (just to keep track of it)

File: projects/storefrontstyles/scss/_components.scss [hunk 9]
```diff
@@ -13,7 +13,6 @@ $cart-components-whitelist: () !default;
 $checkout-components-whitelist: () !default;
 $content-components-whitelist: () !default;
 $misc-components-whitelist: () !default;
-$asm-components-whitelist: () !default;
```
Comment (dunqan) [inline projects/storefrontstyles/scss/_components.scss]: I'm afraid we also need to keep it as is for backward compatibility

File: feature-libs/asm/components/index.ts [hunk 10]
```diff
@@ -0,0 +1,9 @@
+/** we currently do not export all those components and services in the public API */
```
Comment (dunqan) [inline feature-libs/asm/components/index.ts]: Can you double-check if we still want to keep it like this?

File: projects/core/src/process/index.ts [hunk 11]
```diff
@@ -1,3 +1,4 @@
 export * from './process.module';
 export * from './store/process-state';
+export * from './store/reducers/index';
```
Comment (dunqan) [inline projects/core/src/process/index.ts]: do we need it in public api?

File: feature-libs/asm/package.json [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+    "@types/shelljs": "^0.8.7",
+    "jest": "^26.6.2",
+    "rimraf": "^3.0.2",
+    "shelljs": "^0.8.3",
+    "ts-jest": "^26.4.3",
+    "typescript": "~4.0.2"
+  },
+  "dependencies": {
+    "tslib": "^2.0.0"
+  }
```
Comment (plabadie) [inline feature-libs/asm/package.json:37]: I have no hits if I search for `tslib` in asm.  Is it used?

File: feature-libs/asm/package.json [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+    "@spartacus/storefront": "3.0.0-next.6",
+    "@spartacus/schematics": "3.0.0-next.6"
+  },
+  "devDependencies": {
+    "@types/jest": "^26.0.15",
+    "@types/node": "^12.11.1",
+    "@types/shelljs": "^0.8.7",
+    "jest": "^26.6.2",
+    "rimraf": "^3.0.2",
+    "shelljs": "^0.8.3",
```
Comment (plabadie) [inline feature-libs/asm/package.json]: I have no hits if I search for `shelljs` in asm. Is it used?

File: projects/assets/src/translations/en/index.ts (hunk 6 above)
Comment (plabadie) [reply projects/assets/src/translations/en/index.ts]: Done.

File: projects/core/src/asm/config/asm-config.ts (hunk 5 above)
Comment (plabadie) [reply projects/core/src/asm/config/asm-config.ts]: Done.

File: projects/storefrontstyles/scss/_components.scss (hunk 9 above)
Comment (plabadie) [reply projects/storefrontstyles/scss/_components.scss]: Fixed

File: feature-libs/asm/components/index.ts (hunk 10 above)
Comment (plabadie) [reply feature-libs/asm/components/index.ts]: We could add them to the public api now.  I will take care of it.

File: projects/storefrontapp/src/environments/b2c/b2c.feature.ts [hunk 14]
```diff
@@ -30,12 +36,23 @@ export const b2cFeature: FeatureEnvironment = {
...
     }),
+    ConfigModule.withConfig({
+      i18n: {
+        resources: asmTranslations,
+        chunks: asmTranslationChunksConfig,
+        fallbackLang: 'en',
+      },
+    }),
     StoreFinderRootModule,
+    AsmRootModule,
```
Comment (plabadie) [inline projects/storefrontapp/src/environments/b2c/b2c.feature.ts]: We likely need to revert all the changes in this file when we merge the new import structure.

File: feature-libs/asm/package.json [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+    "spartacus",
+    "framework",
+    "storefront",
+    "asm"
+  ],
+  "scripts": {
+    "prepare:schematics": "yarn install",
+    "clean:schematics": "yarn prepare:schematics && rimraf \"schematics/**/*.js\" \"schematics/**/*.js.map\" \"schematics/**/*.d.ts\"",
+    "build:schematics": "yarn clean:schematics && tsc -p ./tsconfig.schematics.json",
+    "test:schematics": "yarn --cwd ../../projects/schematics/ run clean && yarn clean:schematics && jest --config ./jest.schematics.config.js"
```
Comment (plabadie) [inline feature-libs/asm/package.json]: We need to disable schematics temporarily until the schematics are update to support the new storefront app module import structure.

File: projects/core/src/process/index.ts (hunk 11 above)
Comment (plabadie) [reply projects/core/src/process/index.ts]: This looks like a mistake.  I'll remove it.

File: feature-libs/asm/components/index.ts (hunk 10 above)
Comment (plabadie) [reply feature-libs/asm/components/index.ts]: Done: https://github.com/SAP/spartacus/pull/10431/commits/c97a618bcf5fffa4da0eb8e451fe90fc7ce7259b

File: projects/storefrontlib/src/cms-components/cms-lib.module.ts (hunk 7 above)
Comment (plabadie) [reply projects/storefrontlib/src/cms-components/cms-lib.module.ts]: Done

File: projects/storefrontlib/src/recipes/storefront.module.ts (hunk 8 above)
Comment (plabadie) [reply projects/storefrontlib/src/recipes/storefront.module.ts]: Done

**Review by dunqan — APPROVED** [review]

**Pull request merged**